node/session.cpp

```diff
@@ -15,182 +15,256 @@
 #include <node/session.hpp>
 #include <sstream>
 
-namespace
+
+using namespace bzn;
+
+session::session(
+        std::shared_ptr<bzn::asio::io_context_base> io_context,
+        bzn::session_id session_id,
+        boost::asio::ip::tcp::endpoint ep,
+        std::shared_ptr<bzn::chaos_base> chaos,
+        bzn::protobuf_handler proto_handler,
+        std::chrono::milliseconds ws_idle_timeout
+)
+        : session_id(session_id)
+        , ep(std::move(ep))
+        , io_context(std::move(io_context))
+        , chaos(std::move(chaos))
+        , proto_handler(std::move(proto_handler))
+        , idle_timer(this->io_context->make_unique_steady_timer())
+        , ws_idle_timeout(std::move(ws_idle_timeout))
+        , write_buffer(nullptr, 0)
 {
-    const std::chrono::seconds DEFAULT_WS_TIMEOUT_MS{10};
+    LOG(debug) << "creating session " << std::to_string(session_id);
 }
 
+void
+session::start_idle_timeout()
+{
+    this->activity = false;
 
-using namespace bzn;
+    this->idle_timer->expires_from_now(this->ws_idle_timeout);
+    this->idle_timer->async_wait(
```

Comment:
This is going to mess up pub/sub. I originally had this working, but the client never excepts the connection to close when it is waiting for subscribed data.

Reply:
Add a callback to node indicating that this session is going down instead.